proposals/0007-live-edge.md

```diff
@@ -0,0 +1,196 @@
+- Seekable Live Edge = Advertised Live Edge - (`#EXT-X-SERVER-CONTROL:PART-HOLD-BACK` value)
+
+__Context__:
+
+> PART-HOLD-BACK
+> 
+> The value is a decimal-floating-point number of seconds that indicates the server-recommended minimum distance from the end of the Playlist at which clients should begin to play or to which they should seek when playing in Low-Latency Mode. Its value MUST be at least twice the Part Target Duration. Its value SHOULD be at least three times the Part Target Duration. If different Renditions have different Part Target Durations then PART-HOLD-BACK SHOULD be at least three times the maximum Part Target Duration.
+
+  \- From: https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-12#section-4.4.3.8
+
+### Recommended computation of Live Edge - ISO/IEC 23009-1 (aka "MPEG-DASH")
+
+1. __"Standard Latency" Live__
+
+- Let Advertised Live Edge be the computed latest presentation time / duration available as derived from the `MPD` at time T, using `MPD@availabilityStartTime` for reference, and presuming an NTP time-synchronized client and server.
+  - NOTE: Since there are many ways to represent `AdaptationSet`s and their segments, describing the various ways of computing the Advertised Live Edge will be treated as out of scope for this proposal.
+- Seekable Live Edge = Advertised Live Edge - (`MPD@suggestedPresentationDelay` value ?? client player specific value ?? 0)
+
+__Context__:
+
+MPD@suggestedPresentationDelay: 
+> it specifies a fixed delay offset in time from the presentation time of each access unit that is suggested to be used for presentation of each access unit... When not specified, then no value is provided and the client is expected to choose a suitable value.
+
+  \- From: *§5.3.1.2 Table 3 - Semantics of MPD element*
+
+
+2. __Low Latency Live__
+
+- Let Reference Time be derived from the `ProducerReferenceTime` values which correlate to the presentation timeline and wallclock times 
+  - NOTE: Since there are many ways to represent `AdaptationSet`s and their segments, describing the various ways of computing these values will be treated as out of scope for this proposal.
+- Seekable Live Edge = Reference Time - (`ServiceDescription` -> `Latency@target` value)
+
+__Context__:
+
+> The service provider's preferred presentation latency in milliseconds compared to the producer reference time. Indicates a content provider's desire for the content to be presented as close to the indicated latency as is possible given the player's capabilities and observations.
+>
+> This attribute may express latency that is only achievable by low-latency players under favourable network conditions.
+
+  \- From: *Annex K, §K.3.2 Table K.1 - Service Latency* (See Also: *Table K.6 — Semantics of Latency element* from the same section)
+
+## Property: `liveEdgeOffset`
```

Comment:
@luwes 

> I'm having a little trouble with the naming in that it doesn't really say what it is, it does show it's related to the live edge and it's an offset from that. I wish we could somehow incorporate what it's job is. just gonna throw out a stupid name to show what I mean. `livePlayPosition` or `liveTargetTime`.
> 
> _edit: I know it's absolute vs relative but still maybe something more to it's purpose_
> 
> Or am I misinterpreting this?
> 

Agree with the general sentiment that the name "underrepresents its purpose" / has some inherent ambiguity. I'm definitely amenable to some alternatives here. For some context, it was originally named `liveWindowOffset` (also not great, but arguably better), but this creates confusion with another proposed extension property defined in the "DVR State" proposal: `targetLiveWindow`, which @heff and I felt introduced greater confusion, since "window" is being used in two different senses in these two different proposals.

You can find additional conversation from the issue #6 starting at the "Open Questions" and then reading the comments.

> Also for me at least, "holdback" or "delay" is more informative than "offset". In theory an offset could go beyond the live edge in the future

This part may be a misunderstanding. "holdback"/"delay" from the relevant HAS specs is what defines the distinction between the "Advertised Live Edge" vs. the "Seekable Live Edge" (as defined in this proposal). The `liveEdgeOffset`, on the other hand, is what defines the "Live Edge Window" (along with the constrained `seekable.end()`), as defined in this proposal. Think of it as the "fudge factor" for whether or not e.g. the `currentTime` should be treated as "playing live" or not.

Regardless of any name changes to the property, I think your comments suggest there should be some additional informative context added to both the "Guide-level explanation" and the "Reference-level explanation". I'll work on that and see if you think it helps avoid some of the confusion once it's pushed.

Reply:
Raising it here since it's related, in #6 I mentioned [an alt approach](https://github.com/video-dev/media-ui-extensions/issues/6#issuecomment-1421262721) of `liveEdgeStart`.

Reply:
sorry @heff, I missed the last messages in #6. I'll continue the discussion there.

Reply:
Let's keep the discussion here to try to conform to the intended process laid out by @gkatsev.

Reply:
I've updated/added some stuff to hopefully add some clarity around the different values. This includes a section added to the "Guilde-level Explanation" titled "Diagram with HLS reference values for context" (easiest to read here: https://github.com/video-dev/media-ui-extensions/pull/7/files?short_path=6415912#diff-6415912cbdb551127eb5975514c274cb87904befd9ca77ec25808f682ab492d7). Let me know if that helps with clarification of the different parts.

Reply:
> Raising it here since it's related, in https://github.com/video-dev/media-ui-extensions/issues/6 I mentioned https://github.com/video-dev/media-ui-extensions/issues/6#issuecomment-1421262721 of liveEdgeStart.

@heff Yeah that would definitely simplify things programmatically and likely conceptually as well. I'm on board but would like to hear from @gkatsev and @luwes as well.

The tl;dr of this approach:

- Seekable Live - `[seekable.start(0), seekable.end(seekable.length - 1)]` (where `seekable.end()` is constrained per this proposal)
- Live Edge Window - `[liveEdgeStart, seekable.end(seekable.length - 1)]`

Reply:
If I don't hear back from folks by tomorrow afternoon, I'll plan on migrating the proposal to use `liveEdgeStart` with the implications on what the value represents.